subsys/net/l2/openthread/openthread.c

```diff
@@ -334,9 +355,31 @@ static int openthread_init(struct net_if *iface)
 	return 0;
 }
 
+void openthread_joiner_autostart(struct net_if *iface)
+{
+	struct openthread_context *ot_context = net_if_l2_data(iface);
+	otError error;
+
+	error = otJoinerStart(ot_context->instance, CONFIG_OPENTHREAD_JOINER_PSKD, NULL, PACKAGE_NAME,
+			CONFIG_OPENTHREAD_PLATFORM_INFO, PACKAGE_VERSION, NULL, &ot_joiner_start_handler, ot_context);
+
+	switch (error) {
+	case OT_ERROR_NONE:
+		NET_INFO("Start joiner success");
+		break;
+	default:
+		NET_ERR("Faild to start joiner [%d]", error);
+		break;
+	}
+}
+
 void ieee802154_init(struct net_if *iface)
 {
 	openthread_init(iface);
+
+#if defined(CONFIG_OPENTHREAD_JOINER_AUTOSTART)
```

Comment:
I'm afraid the logic here is too simple. With this approach, application will try to initiate the commissioning on each reboot.

There's an API function to verify that , check https://github.com/openthread/openthread/blob/bd41c63a0a041a4a43a5deab421d719a7011d5ca/include/openthread/dataset.h#L275

In case there's a valid network stored, you'd only need to start it with `otThreadSetEnabled`, otherwise initiate the joining process.

And I'd move this piece within `openthread_init` to keep the network starting logic in a single place.

Reply:
Thanks for your reply!

Which do you mean:

```
otThreadSetEnabled(..., true);
if (!otDatasetIsCommissioned()) {
  otThreadSetEnabled(..., false);
  otJoinerStart(...);
}

callback() {
  otThreadSetEnabled(..., true);
}
```

or,

```
if (!otDatasetIsCommissioned()) {
  otJoinerStart(...);
}

callback() {
  otThreadSetEnabled(..., true);
}
```
?
Should I enable thread absolutely before call otDatasetIsCommissioned() to get network validation?

Reply:
I don't understand why otDatasetIsCommissioned() returns false absolutely even zephyr joined successful and rebooted or not in my environment...

Reply:
I meant something like this:
```
if (!otDatasetIsCommissioned()) {
    otJoinerStart(...);
} else {
    otThreadSetEnabled(..., true);
}

callback() {
    otThreadSetEnabled(..., true);
}
```

`otThreadSetEnabled(..., true);` simply starts the Thread network. It will use the parameters that we obtained during commissionning process (that's why we call it in the joiner callback). But the same parameters are (or at least should, if not we have a bug somewhere) stored in the flash, hence after reboot there's no need to go through the commissioning process again, we can start rightaway.

> And I don't understand why otDatasetIsCommissioned() returns false absolutely even zephyr joined successful and rebooted or not in my environment...

Hmm that's weird, let my try your PR on desk setup, perhaps there's something fishy going on. What hardware do you use?

Reply:
I applied changes you told and added a patch for test:
https://github.com/takumiando/zephyr/commits/openthread_autostart_joiner_test

I'm using nrf52840_pca10056 board, and use sample/net/socket/echo_client with test patch.

```
$ cd sample/net/socket/echo_client
$ mkdir build && cd build
$ cmake -GNinja -DBOARD=nrf52840_pca10056 -DOVERLAY_CONFIG=overlay-ot.conf ..
$ ninja
```

And run it, my board joins to my network:

```
***** Booting Zephyr OS v1.14.0-rc1-625-g3a012c9502e4 *****
[00:00:14.357,666] <inf> net_l2_openthread: State changed! Flags: 0x00000010 Current role: 0
[00:00:14.451,110] <inf> net_l2_openthread: Join success
[00:00:14.454,010] <inf> net_l2_openthread: State changed! Flags: 0x0000903f Current role: 1
[00:00:16.446,075] <inf> net_l2_openthread: State changed! Flags: 0x00000040 Current role: 1
[00:00:17.412,109] <inf> net_l2_openthread: State changed! Flags: 0x000012a4 Current role: 2
```

Then, I reset my board but it tries to start joiner again...

```
***** Booting Zephyr OS v1.14.0-rc1-625-g3a012c9502e4 *****
[00:00:17.249,816] <inf> net_l2_openthread: State changed! Flags: 0x00000010 Current role: 0
[00:00:17.345,092] <inf> net_l2_openthread: Join success
[00:00:17.348,052] <inf> net_l2_openthread: State changed! Flags: 0x0000903f Current role: 1
[00:00:19.340,057] <inf> net_l2_openthread: State changed! Flags: 0x00000040 Current role: 1
[00:00:20.800,231] <inf> net_l2_openthread: State changed! Flags: 0x000012a4 Current role: 2
```

My understanding is otDatasetIsCommissioned() returns false both these cases.
I can't understand why it does... 

I dumped and compared flash on my board by nrfjprog at before joined and after.

```
--- before_join	2019-03-01 11:12:26.249137691 +0900
+++ after_join	2019-03-01 11:13:10.633884614 +0900
@@ -64766,20 +64766,20 @@
 0x000FCFD0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
 0x000FCFE0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
 0x000FCFF0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD000: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD010: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD020: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD030: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD040: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD050: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD060: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD070: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD080: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD090: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD0A0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD0B0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD0C0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
-0x000FD0D0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
+0x000FD000: C5EE BE5C 0003 00F4 0024 0000 0000 0000   |..\.....$.......|
+0x000FD010: 0001 0000 03E8 0000 03E8 0000 0000 0000   |................|
+0x000FD020: 0000 0000 0000 0000 0000 0000 0000 0000   |................|
+0x000FD030: 0001 00F4 005E CE82 0300 0000 3514 0006   |....^........5..|
+0x000FD040: 0004 FF1F 02E0 0508 5542 A7AF 12E8 0767   |........BU....g.|
+0x000FD050: FD08 4205 AF55 00A7 0500 DE10 2B7B 789D   |...BU.......{+.x|
+0x000FD060: D48D 1A57 F3E2 A543 0E4D 03F6 6404 6765   |..W...C.M....deg|
+0x000FD070: 0175 AB02 04CD 0010 0000 0000 0000 0000   |u...............|
+0x000FD080: 0000 0000 0000 0C00 0203 FFA0 080E 0000   |................|
+0x000FD090: 0000 0000 0000 FFFF 0004 00F4 0008 724C   |..............Lr|
+0x000FD0A0: F456 724C 1925 A668 0003 00F4 0024 724C   |V.Lr%.h.....$.Lr|
+0x000FD0B0: 0F02 9C1E 0001 0000 03EA 0000 03E8 0000   |................|
+0x000FD0C0: 9A22 5306 5866 70AA 8A27 92E9 05B2 97AD   |"..SfX.p'.......|
+0x000FD0D0: F91C 6A10 FFFF FFFF FFFF FFFF FFFF FFFF   |...j............|
 0x000FD0E0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
 0x000FD0F0: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
 0x000FD100: FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF   |................|
```

otDatasetIsCommissioned() checks this difference region, is that right?

Reply:
Hi @takumiando , thanks for investigating! `otDatasetIsCommissioned` checks the Active Dataset loaded into RAM from flash memory during OpenThread instance initialization. I did some debugging on my own, and realized that `otDatasetIsCommissioned` returns true just after `otInstanceInitSingle`. Yet, when called just before joiner start, it reutrns false. Therefore it seems that either something corrupts the dataset in RAM, or it is discarded for whatever reason and overwritten. I"m going to investigate the issue further and hopefully propose a patch to fix it.

Reply:
Ok, found the showstopper.

> or it is discarded for whatever reason and overwritten

Apparently that was the case. I missed the fact, that network name is also obtained during commissioning, so we should not set it manually with `otThreadSetNetworkName`. Otherwise, the dataset is invalidated:
https://github.com/openthread/openthread/blob/8c13df5a51f4b0428d73f0bc57a01506c178e421/include/openthread/thread.h#L459

Therefore, bundling `otThreadSetNetworkName` with other parameters underneath 
`#if !defined(CONFIG_OPENTHREAD_JOINER_AUTOSTART)` solves the issue! Just note, that OT flash data region is not cleared when you program the application - you may want to erase the board with `nrfjprog -e` to test the joining procedure.

Reply:
Thanks for investigating!
I confirmed that zephyr does not start joiner after reboot.

I fixed this issue and force-pushed to the same branch!